### ISSUE TITLE
Problems with q-model.kanachan in simulation and mjai.app in new version

### ISSUE THREAD

**Issue by KimamanaNeko** [Cryolite/kanachan#44]
How to make a prediction based on a new version of q-model.
I loaded q-model.kanachan with load_model.
But it seems I can't pass canidates to Value Network V

prediction = self.__q_model(sparse, numeric, progression, candidates_) 
encode = self.encoder(sparse, numeric, progression, candidates)      ...iql/value_model.py", line 67, in forward 
q1, _ = self.qv1_model(sparse, numeric, progression, candidates)     ...ilql/q_model.py", line 25, in forward  
return self.decoder(candidates, encode)                                              ...ilql/qv_model.py", line 91, in forward  
value: torch.Tensor = self.value_decoder(candidates, encode)           ...ilql/qv_model.py", line 60, in forward  
assert value.size(0) == candidates.size(0)                                           ...iql/value_model.py", line 50, in forward  
AssertionError

Btw bert's model.kanachan works fine for me.
Any tips would be greatly appreciated!

**Comment by KimamanaNeko** [comment]
I did some more research.
I found that in value[mask] mask is all false

This could be due to incorrect input data format ,
My input data is exactly the same as in bert's model.
`prediction = self.__q_model(sparse, numeric, progression, candidates_)`
Can you give me a little hint about the input format of the q-model?

**Comment by KimamanaNeko** [comment]
I thinks need`mask = (candidates == NUM_TYPES_OF_ACTIONS+1)`

**Comment by KimamanaNeko** [comment]
Other information
I confirmed the predictions of ilql's model
the encoder in the forward of QVModel(nn.Module) is
`tensor([[[-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002],
         [-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002],
         [-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002],
         ...,
         [-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002],
         [-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002],
         [-0.0002,  0.0001,  0.0005,  ..., -0.0736, -0.0005, -0.0002]]])`

**Comment by KimamanaNeko** [comment]
candidates tensor([[ 28,  32,  36,  44,  46,  64,  88,  92, 120, 124, 144, 547, 547, 547,
         547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547,
         547, 547, 547, 547]], dtype=torch.int32)
NUM_TYPES_OF_ACTIONS 546

mask = (candidates == NUM_TYPES_OF_ACTIONS)
This could be a mistake.

**Comment by Cryolite** [comment]
Could you tell me the commit hash of the `kanachan` repository where the error is occurring? You can find out the commit hash by executing `git show --format='%H' --no-patch`. This information might help me understand the cause of the problem. Also, if you have made local changes to the `kanachan` repository, could you provide the results of the `git diff` command?

**Comment by KimamanaNeko** [comment]
`fd1303377be509b3c430eb1c66e2037f5e95d4c9`
The model training was done using your git main branch, without any modifications.
load the ilql model and perform the prediction test in mjai.app.

class Kanachan:
    def __init__(self) -> None:
        qmodel_snapshot_path = Path('./q-model.kanachan')
        qmodel_state = load_model(
            qmodel_snapshot_path, map_location=torch.device('cpu'))
        self.__model = qmodel_state 

        with open('./game.json', encoding='UTF-8') as f:
            game_config = json.load(f)
        self.__game_state = GameState(**game_config)

        self.__round_state = RoundState()
        
This is all the changes I made to mjai.app to read the ilql model, replacing the old encoder and decoder method.

**Comment by KimamanaNeko** [comment]
Description
When I test the read of the ilql model  to predictions in the latest model format, I get the error `assert value.size(0) == candidates.size(0)`

Steps to Reproduce
Steps to reproduce the behavior:
1.train a ilql model
2.move `q-model.kanachan` to mjai.app
3.use `load_model` to get `nn.Module`
4.`self.__q_model = qmodel_state`

Expected Behavior
Returns the correct predicted value

Screenshots
![image](https://github.com/Cryolite/kanachan/assets/46703388/4a342a21-0cd5-4ad0-be6a-6d565c8c3382)

Environment:
Train Version: main  fd13033 on May 7 `fd1303377be509b3c430eb1c66e2037f5e95d4c9`

Additional Context
In the forward of value_model.py, mask = (candidates == NUM_TYPES_OF_ACTIONS)

candidates 
tensor([[ 28, 32, 36, 44, 46, 64, 88, 92, 120, 124, 144, 547, 547, 547,
547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547,
547, 547, 547, 547]], dtype=torch.int32)

NUM_TYPES_OF_ACTIONS 546

Maybe this `mask` isn't working as well as it should
value = value[mask] Always output a null value

**Comment by Cryolite** [comment]
When making predictions with the current version of the model, modifications to the padding of `candidates` are necessary. The previous padding specification for `candidates` was simply to fill with `547` (`NUM_TYPES_OF_ACTIONS + 1`). The current padding specification for `candidates` has been changed to first fill with just one instance of `546` (`NUM_TYPES_OF_ACTIONS`), and then fill the rest with `547`. Therefore, the correct example of `candidates` is `tensor([[ 28, 32, 36, 44, 46, 64, 88, 92, 120, 124, 144, 546, 547, 547,
547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547, 547,
547, 547, 547, 547]], dtype=torch.int32)` (please note that `546` is at the beginning of the padding). This change has already been reflected in the `develop` branch's simulation and mjai.app.
